xdem/volume.py

```diff
@@ -506,3 +507,194 @@ def local_hypsometric_interpolation(voided_ddem: Union[np.ndarray, np.ma.masked_
+    ddem_arr, ddem_nans = xdem.spatial_tools.get_array_and_mask(voided_ddem.squeeze())
+    ref_arr, ref_nans = xdem.spatial_tools.get_array_and_mask(ref_dem.squeeze())
+
+    # The reference DEM should be void free
+    assert np.count_nonzero(ref_nans) == 0, "Reference DEM has voids"
+
+    # If the regional signal was not given as an argument, find it from the dDEM.
+    if regional_signal is None:
+        regional_signal = get_regional_hypsometric_signal(
+            ddem=ddem_arr,
+            ref_dem=ref_arr,
+            glacier_index_map=glacier_index_map
+        )
+
+    # The unique indices are the unique glaciers.
+    unique_indices = np.unique(glacier_index_map)
+
+    # Make a copy of the dDEM which will be filled iteratively.
+    ddem_filled = ddem_arr.copy()
+    # Loop over all glaciers and fill the dDEM accordingly.
+    for i in unique_indices:
+        if i == 0:  # i==0 is assumed to mean stable ground.
+            continue
+        # Create a mask representing a particular glacier.
+        glacier_values = (glacier_index_map == i)
+
+        # The inlier mask is where that particular glacier is and where nans don't exist.
+        inlier_mask = glacier_values & ~ddem_nans
+
+        # If the fractional coverage is smaller than the given threshold, skip the glacier.
+        if (np.count_nonzero(inlier_mask) / np.count_nonzero(glacier_values)) < min_coverage:
+            continue
+
+        # Extract only the finite difference and elevation values that correspond to the glacier.
+        differences = ddem_arr[inlier_mask]
+        elevations = ref_arr[inlier_mask]
+
+        # Get the reference elevation min and max
+        elev_min = ref_arr[glacier_values].min()
+        elev_max = ref_arr[glacier_values].max()
+
+        # Copy the signal
+        signal = regional_signal["w_mean"].copy()
+        # Scale the signal elevation midpoints to the glacier elevation range.
+        midpoints = signal.index.mid
+        midpoints *= elev_max - elev_min
+        midpoints += elev_min
+        step = midpoints[1] - midpoints[0]
+        # Create an interval structure from the midpoints and the step size.
+        signal.index = pd.IntervalIndex.from_arrays(left=midpoints - step / 2, right=midpoints + step / 2)
+
+        # Find the hypsometric bins of the glacier.
+        hypsometric_bins = hypsometric_binning(
+            ddem=differences,
+            ref_dem=elevations,
+            bins=np.r_[[signal.index.left[0]], signal.index.right],  # This will generate the same steps as the signal.
+            kind="custom"
+        )
+        # Check which of the bins were non-empty.
+        non_empty_bins = np.isfinite(hypsometric_bins["value"])
+
+        # A theoretical minimum of two bins are needed for the rest of the analysis, but that's a bit low, right?
+        # I'll set this to 4 right now, and we can discuss this.
+        if np.count_nonzero(non_empty_bins) < 4:
```

Comment:
Uh, could you elaborate on what you mean here? I think I see the problem but not really the solution!

Reply:
I was thinking that using an arbitrary percentage of the number of bins (instead of an arbitary number of bins) would be a bit better. For now `nbins` is hardcoded as 20 so choosing 4 bins corresponds to 20% of elevation range bins, but we might want to change this parameter in the future!

Reply:
Added as an argument in a8d8747!